rescript/get_gtdb.py

```diff
@@ -0,0 +1,138 @@
+
+
+def get_gtdb_data(ctx, version='207', domain='Both'):
+
+    ver_dom_dict = defaultdict(lambda: defaultdict(dict))
+
+    # Subset dict if needed, but keep same structure
+    # Although we can run the following merge actions on a list of one
+    # i.e. 'Archaea', we do not want to confuse anyone when looking
+    # at provenance, by running a merge command for no reason.
+    if domain == 'Both':
+        ver_dom_dict[version] = VERSION_MAP_DICT[version]
+    else:
+        ver_dom_dict[version][domain] = VERSION_MAP_DICT[version][domain]
+
+    queries = _assemble_queries(ver_dom_dict)
+    tax_q, seqs_q = _retrieve_data_from_gtdb(queries)
+
+    if domain == 'Both':
+        merge_gtdb_seqs = ctx.get_action('feature_table', 'merge_seqs')
+        merge_gtdb_taxonomy = ctx.get_action('feature_table', 'merge_taxa')
+        print('\n  Merging taxonomy data...')
+        merged_gtdb_tax, = merge_gtdb_taxonomy(data=tax_q)
+        print('\n  Merging sequence data...')
+        merged_gtdb_seqs, = merge_gtdb_seqs(data=seqs_q)
+    else:
+        merged_gtdb_tax = tax_q[0]
+        merged_gtdb_seqs = seqs_q[0]
+
+    print('\n Saving files...\n')
+    return merged_gtdb_tax, merged_gtdb_seqs
+
+
+def _assemble_queries(ver_dom_dict):
+    queries = defaultdict(lambda: list())
+
+    base_seq_url = ('https://data.gtdb.ecogenomic.org/releases/release%s/'
+                    '%s.0/genomic_files_reps/%s_ssu_reps_r%s.tar.gz')
+    base_tax_url = ('https://data.gtdb.ecogenomic.org/releases/release%s/'
+                    '%s.0/%s_taxonomy_r%s.tsv.gz')
+
+    for version, dcp in ver_dom_dict.items():
+        for dom, cp in dcp.items():
+
+            queries['Taxonomy'].append(
+                (dom,
+                 base_tax_url % (version, version, cp, version),
+                 'FeatureData[Taxonomy]',
+                 'HeaderlessTSVTaxonomyFormat'))
+
+            queries['Sequence'].append(
+               (dom,
+                base_seq_url % (version, version, cp, version),
+                'FeatureData[Sequence]',
+                'DNAFASTAFormat'))
+    return queries
+
+
+def _retrieve_data_from_gtdb(queries):
+    '''
+    Download data from gtdb, given a list of queries.
+
+    queries: {'Taxonomy':(domain, url, type, format), (...),
+              'Sequence':(domain, url, type, format), (...)}
+    '''
+
+    tax_results = []
+    seq_results = []
+
+    print('\nDownloading and processing raw files ... \n')
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        for sttype, q_info in queries.items():
+            for domain, url, dtype, fmt in q_info:
+                print('Retrieving {0} for {1} from {2}'.format(
+                      sttype, domain, url))
+                # grab url
+                bn = os.path.basename(url)
+                destination = os.path.join(tmpdirname, bn)
+                try:
+                    urlretrieve(url, destination)
```

Comment:
I agree with @nbokulich that no data should be fetched in the tests. If you refactor a little (see my other comment below) and then only mock this `urlretrieve` function, it should be quite straightforward to test this entire function :)